lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -197,17 +213,54 @@ void SPIRVRegularizeLLVM::buildFunnelShiftLeftFunc(Function *FSHLFunc) {
   Builder.CreateRet(FunnelShiftRes);
 }
 
-void SPIRVRegularizeLLVM::lowerFunnelShiftLeft(IntrinsicInst *FSHLIntrinsic) {
+void SPIRVRegularizeLLVM::buildFunnelShiftRightFunc(Function *FSHRFunc) {
+  if (!FSHRFunc->empty())
+    return;
+
+  auto *RotateBB = BasicBlock::Create(M->getContext(), "rotate", FSHRFunc);
+  IRBuilder<> Builder(RotateBB);
+  Type *Ty = FSHRFunc->getReturnType();
+  // Build the actual funnel shift rotate logic.
+  // In the comments, "int" is used interchangeably with "vector of int
+  // elements".
+  FixedVectorType *VectorTy = dyn_cast<FixedVectorType>(Ty);
+  Type *IntTy = VectorTy ? VectorTy->getElementType() : Ty;
+  unsigned BitWidth = IntTy->getIntegerBitWidth();
+  ConstantInt *BitWidthConstant = Builder.getInt({BitWidth, BitWidth});
+  Value *BitWidthForInsts =
+      VectorTy ? Builder.CreateVectorSplat(VectorTy->getNumElements(),
+                                           BitWidthConstant)
+               : BitWidthConstant;
+  auto *RotateModVal =
+      Builder.CreateURem(/*Rotate*/ FSHRFunc->getArg(2), BitWidthForInsts);
+  // Shift the less significant number right, the "rotate" number of bits
+  // will be 0-filled on the left as a result of this regular shift.
+  auto *ShiftRight = Builder.CreateLShr(FSHRFunc->getArg(0), RotateModVal);
+  // We want the "rotate" number of the second int's LSBs to occupy the
```

Comment:
Now that I'm re-reading this, we may be even better off with:
```suggestion
  // We want the "rotate" number of the more significant int's LSBs to occupy the
```
It's natural for Westerners to count from left to right, therefore part of the potential readers might confuse "the second int" for "the rightmost int" instead of simply "the other int that we haven't touched yet".